### PR TITLE
Preserve graph state

### PR DESCRIPTION
**Pull request by whiskeyjay**

### CODE REVIEW

File: src/graph/GraphViewServer.ts [hunk 1]
```diff
@@ -22,6 +22,10 @@ export class GraphViewServer extends EventEmitter {
   private _httpServer: http.Server;
   private _port: number | undefined;
   private _socket: SocketIO.Socket;
+  private _lastQuery: string | undefined;
```
Comment (StephenWeatherford) [inline src/graph/GraphViewServer.ts]: Can we turn these into a single object with members?  Something like
interface ClientViewState {
  query?: string;
  data?: any[];
  errorMessage?: string
}

File: src/graph/client/graphClient.ts [hunk 2]
```diff
@@ -131,14 +156,24 @@ export class GraphClient {
...
+  }
+
+  public setQuery(query: string) {
+    this.emitToHost('setQuery', query);
+  }
+
+  private setView() {
+    htmlElements.graphRadio.checked = this._graphView;
+    htmlElements.jsonRadio.checked = !this._graphView;
+    d3.select(htmlElements.graphSection).classed("active", this._graphView);
```
Comment (StephenWeatherford) [inline src/graph/client/graphClient.ts]: nit: !! would make it more understandable

File: src/graph/GraphViewServer.ts [hunk 3]
```diff
@@ -153,6 +162,24 @@ export class GraphViewServer extends EventEmitter {
...
+    console.log('getPageState');
+
+    if (this._lastQuery) {
+      this._socket.emit('setPageState', this._lastQuery, this._lastErrorMsg, this._lastResults, this._lastView);
+    }
+  }
+
+  private handleSetQuery(query: string) {
+    console.log('setQuery');
+    this._lastQuery = query;
```
Comment (StephenWeatherford) [inline src/graph/GraphViewServer.ts]: This means that if you query for results, wait for the results to show, then type a new query (or the start of a new query), but don't hit Execute, then tab away and back, the new query will show, and the old results will animate in, making it look like they're the new results.  Correct?

Perhaps you should throw away the last results (in the server, not the client) when the query changes?

File: src/graph/client/graphClient.ts [hunk 4]
```diff
@@ -96,6 +102,21 @@ export class GraphClient {
...
+      if (!errorMsg) {
+        this.showResults(results);
+
+        if (view === 'json') {
+          this.selectJsonView();
+        } else {
+          this.selectGraphView();
+        }
+      } else {
+        this.setStateError(errorMsg);
```
Comment (StephenWeatherford) [inline src/graph/client/graphClient.ts]: You should still set the view back to what the user had last selected, even if there's an error.

File: src/graph/client/graphClient.ts [hunk 5]
```diff
@@ -122,6 +143,10 @@ export class GraphClient {
     });
   }
 
+  public getPageState() {
+    this.emitToHost('state');
```
Comment (StephenWeatherford) [inline src/graph/client/graphClient.ts]: 'getPageState'

File: src/graph/client/graphClient.ts [hunk 6]
```diff
@@ -131,14 +156,24 @@ export class GraphClient {
...
-    d3.select(htmlElements.jsonSection).classed("active", true);
+    this.setView();
+  }
+
+  public setQuery(query: string) {
+    this.emitToHost('setQuery', query);
+  }
+
+  private setView() {
+    htmlElements.graphRadio.checked = this._graphView;
```
Comment (StephenWeatherford) [inline src/graph/client/graphClient.ts]: You should call selectGraphView or selectJsonView() instead.

Comment (whiskeyjay) [reply src/graph/client/graphClient.ts]: Both select view functions call this.

File: resources/graphClient/graphClient.css [hunk 7]
```diff
@@ -172,6 +172,11 @@ textarea {
     padding-top: 18%;
 }
 
+#stats {
```
Comment (StephenWeatherford) [inline resources/graphClient/graphClient.css]: Careful on merge, this is not the current version of the code.

File: resources/graphClient/graphClient.html [hunk 8]
```diff
@@ -63,6 +63,7 @@ <h1 id="title">
             <div id="resultsSection">
                 <div id="graphSection" class="active">
                     <!-- SVG will be inserted here -->
+                    <span id="stats"></span>
```
Comment (StephenWeatherford) [inline resources/graphClient/graphClient.html]: Careful on merge, this is not the current version of the code.

File: src/graph/GraphViewServer.ts [hunk 9]
```diff
@@ -22,9 +22,22 @@ export class GraphViewServer extends EventEmitter {
   private _httpServer: http.Server;
   private _port: number | undefined;
   private _socket: SocketIO.Socket;
+  private _previousPageState: {
+    lastQuery: string | undefined,
```
Comment (StephenWeatherford) [inline src/graph/GraphViewServer.ts]: nit: you could remove the "last" from all these

File: src/graph/client/graphClient.ts [hunk 10]
```diff
@@ -15,6 +17,8 @@ declare let d3: any;
 const animationStepMs = 50; // TODO: optimize.  Slow down ticks?
 const graphWidth = 1200, graphHeight = 500; //TODO: be resizable or adapt to editor size
 const defaultQuery = "g.V()";
+const maxNodes = 1000;
```
Comment (StephenWeatherford) [inline src/graph/client/graphClient.ts]: Careful on merge, this is not the current version of the code.

File: src/graph/client/graphClient.ts [hunk 11]
```diff
@@ -122,6 +122,11 @@ export class GraphClient {
     this._socket.on('setPageState', (previousState) => {
       htmlElements.queryInput.value = previousState.lastQuery;
 
+      if (previousState.isQueryRunning) {
```
Comment (StephenWeatherford) [inline src/graph/client/graphClient.ts]: Still need to set the correct tab being active

**Pull request merged**